src/dfx/src/lib/canister_info.rs

```diff
@@ -273,30 +228,46 @@ impl CanisterInfo {
     }
 
     pub fn get_output_idl_path(&self) -> Option<PathBuf> {
-        if let Ok(info) = self.as_info::<MotokoCanisterInfo>() {
-            Some(info.get_output_idl_path().to_path_buf())
-        } else if let Ok(info) = self.as_info::<CustomCanisterInfo>() {
-            Some(info.get_output_idl_path().to_path_buf())
-        } else if let Ok(info) = self.as_info::<AssetsCanisterInfo>() {
-            Some(info.get_output_idl_path().to_path_buf())
-        } else if let Ok(info) = self.as_info::<RustCanisterInfo>() {
-            Some(info.get_output_idl_path().to_path_buf())
-        } else {
-            self.get_extra_optional("candid")
-                .unwrap_or(None)
-                .or_else(|| self.remote_candid.clone())
+        match &self.type_specific {
+            CanisterTypeProperties::Motoko { .. } => self
+                .as_info::<MotokoCanisterInfo>()
+                .map(|x| x.get_output_idl_path().to_path_buf()),
+            CanisterTypeProperties::Custom { .. } => self
+                .as_info::<CustomCanisterInfo>()
+                .map(|x| x.get_output_idl_path().to_path_buf()),
+            CanisterTypeProperties::Assets { .. } => self
+                .as_info::<AssetsCanisterInfo>()
+                .map(|x| x.get_output_idl_path().to_path_buf()),
+            CanisterTypeProperties::Rust { .. } => self
+                .as_info::<RustCanisterInfo>()
+                .map(|x| x.get_output_idl_path().to_path_buf()),
         }
+        .ok()
+        .or_else(|| self.remote_candid.clone())
```

Comment:
Looks like we lost the equivalent of the `self.get_extra_optional("candid")` lookup before falling back to the `remote_candid` field.  This was added recently to handle the case where the type lookup failed.  Maybe @sesi200 will have more insight.

Reply:
This was added as a fallback when no canister type can be inferred ([Jira here](https://dfinity.atlassian.net/browse/SDK-506), PR #2262). Since we now handle this with exhaustive pattern matching and an option will always be chosen it should be fine to remove this branch.

Thinking a bit more I feel that we should not even rely on `get_output_idl_path` as the only way to determine the candid to use in `commands/canister/call.rs`. Using this as a fallback for OUTPUT idl sounds like a bug waiting to happen. If we refactor this to use proper logic (`candid` or `remote/candid` if available, otherwise `output_idl_path`) in `call.rs`, `get_output_idl_path` could get rid of the `.or_else` branch.